File: python/502_IPO_hard.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        n, cur = len(profits), 0
        pq = []
        data = [(capital[i], profits[i]) for i in range(n)]
        data.sort(key = lambda x : x[0])
        for i in range(k):
            while cur < n and data[cur][0] <= w:
                heappush(pq, -data[cur][1])
                cur += 1
            if pq:
                w -= heappop(pq)
            else:
                break
        return w

File: cpp/502_IPO_hard.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n = profits.size();
        int cur = 0;
        vector<pair<int, int>> data;
        priority_queue<int> pq;
        for (int i=0; i<n; i++)
            data.push_back({capital[i], profits[i]});
        sort(data.begin(), data.end());
        for (int i=0; i<k; i++) {
            while (cur < n && data[cur].first <= w)
                pq.push(data[cur++].second);
            if (!pq.empty()) {
                w += pq.top();
                pq.pop();
            }
            else
                break;
        }
        return w;
    }
};